pkg/controller/istiocsr/certificates.go

```diff
@@ -0,0 +1,139 @@
+	key := types.NamespacedName{
+		Name:      desired.GetName(),
+		Namespace: desired.GetNamespace(),
+	}
+	exist, err := r.Exists(r.ctx, key, fetched)
+	if err != nil {
+		return FromClientError(err, "failed to check %s certificate resource already exists", certificateName)
+	}
+
+	if exist && istioCSRCreateRecon {
+		r.eventRecorder.Eventf(istiocsr, corev1.EventTypeWarning, "ResourceAlreadyExists", "%s certificate resource already exists, maybe from previous installation", certificateName)
+	}
+	if exist && hasObjectChanged(desired, fetched) {
+		r.log.V(1).Info("certificate has been modified, updating to desired state", "name", certificateName)
+		if err := r.UpdateWithRetry(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to update %s certificate resource", certificateName)
+		}
+		r.eventRecorder.Eventf(istiocsr, corev1.EventTypeNormal, "Reconciled", "certificate resource %s reconciled back to desired state", certificateName)
+	} else {
+		r.log.V(1).Info("certificate resource already exists and is in expected state", "name", certificateName)
+	}
+	if !exist {
+		if err := r.Create(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to create %s certificate resource", certificateName)
+		}
+		r.eventRecorder.Eventf(istiocsr, corev1.EventTypeNormal, "Reconciled", "certificate resource %s created", certificateName)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) getCertificateObject(istiocsr *v1alpha1.IstioCSR, resourceLabels map[string]string) (*certmanagerv1.Certificate, error) {
+	if istiocsr.Spec.IstioCSRConfig == nil {
+		return nil, fmt.Errorf("not creating certificate resource, %s/%s spec.IstioCSRConfig is empty", istiocsr.GetNamespace(), istiocsr.GetName())
+	}
+
+	certificate := decodeCertificateObjBytes(assets.MustAsset(certificateAssetName))
+
+	updateNamespace(certificate, istiocsr.Spec.IstioCSRConfig.Istio.Namespace)
+	// add custom label for identification on the object created in different namespace.
+	labels := make(map[string]string, len(resourceLabels)+1)
+	for k, v := range resourceLabels {
+		labels[k] = v
+	}
+	labels[istiocsrNamespaceMappingLabelName] = istiocsr.GetNamespace()
+	certificate.SetLabels(labels)
+
+	if err := updateCertificateParams(istiocsr, certificate); err != nil {
+		return nil, NewIrrecoverableError(err, "failed to update certificate resource for %s/%s istiocsr deployment", istiocsr.GetNamespace(), istiocsr.GetName())
+	}
+
+	return certificate, nil
+}
+
+func updateCertificateParams(istiocsr *v1alpha1.IstioCSR, certificate *certmanagerv1.Certificate) error {
+	certificate.Spec.CommonName = istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CommonName
+	if istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CommonName == "" {
+		certificate.Spec.CommonName = fmt.Sprintf(istiodCertificateCommonNameFmt, istiocsr.Spec.IstioCSRConfig.Istio.Namespace)
+	}
+
+	dnsNames := make([]string, len(istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CertificateDNSNames))
```

Comment:
the default behavior here lies that we're adding only the commonName in the dnsNames list i.e. in practice only one dnsName when user specifies nothing, but it has to be:

```diff
   dnsNames
   - istiod.${NAMESPACE}.svc
+  - istiod-basic.${NAMESPACE}.svc
```

for istio to work correctly; fails otherwise.

cc: @PillaiManish

Reply:
By default, the service-mesh-operator installation installs Istio operand that requires the `istiod-basic` as a required dns name in the certificate, do you think for every install users would need to add `basic` in the revisions list as a must?

Reply:
I think this depends on the name given to the SMCP resource. If it's anything other than default, we need to add it to revisions list for adding the required dnsName in the istiod certificate.

Reply:
I see: https://github.com/openshift/cert-manager-operator/pull/226/files#diff-6780e9cfdc6d17cfa06da40b92ceae03337dfda6165d26a035691ce9e1921945R4,

so in case the name of SMCP is `default` instead
this is already taken care (i.e. no need to add extra dnsName in the certificate), right?

Reply:
yeah, correct.